### PR TITLE
Fp16 training

### PR DESCRIPTION
**Pull request by typhoonzero**

### CODE REVIEW

File: python/paddle/fluid/layers/learning_rate_scheduler.py [hunk 1]
```diff
@@ -230,17 +231,17 @@ def polynomial_decay(learning_rate,
         if cycle:
             div_res = ops.ceil(global_step / decay_steps)
             zero_var = tensor.fill_constant(
-                shape=[1], dtype='float32', value=0.0)
+                shape=[1], dtype="float32", value=0.0)
             one_var = tensor.fill_constant(
-                shape=[1], dtype='float32', value=1.0)
+                shape=[1], dtype="float32", value=1.0)
```
Comment (chengduoZH) [inline python/paddle/fluid/layers/learning_rate_scheduler.py]: Why do you change this?

File: python/paddle/fluid/optimizer.py [hunk 2]
```diff
@@ -74,6 +74,12 @@ def _create_global_learning_rate(self):
         lr = self._global_learning_rate()
 
         if isinstance(lr, framework.Variable):
+            # NOTE: cast to optimizer dtype if needed, this is for
+            # training with fp16 which learning rate should also be
+            # a 
+            if self._dtype:
+                if lr.dtype != self._dtype:
+                    layers.cast(lr, self._dtype)
```
Comment (chengduoZH) [inline python/paddle/fluid/optimizer.py]: There doesn't have variable to accept the return of `layers.cast`.

File: python/paddle/fluid/layers/learning_rate_scheduler.py (hunk 1 above)
Comment (typhoonzero) [reply python/paddle/fluid/layers/learning_rate_scheduler.py]: Done.

File: python/paddle/fluid/optimizer.py (hunk 2 above)
Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py]: Removed and will add in next PR

**Review by chengduoZH — APPROVED** [review]

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -2797,6 +2797,10 @@ def batch_norm(input,
     helper = LayerHelper('batch_norm', **locals())
     dtype = helper.input_dtype()
 
+    # use fp32 for bn parameter
+    if dtype == core.VarDesc.VarType.FP16:
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2801]: I don't like this hardcode. For now, we should change the model and add cast op before bn.

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -2797,6 +2797,10 @@ def batch_norm(input,
...
+        dtype = core.VarDesc.VarType.FP32
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2802]: what's the output type, is it fp16 or fp32. What if the next op needs fp16?

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (typhoonzero) [reply python/paddle/fluid/layers/nn.py:2801]: Seems currently all fp16 training should use batch norm with fp32 parameters, which includes moving mean, var and trainable scale, bias

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (typhoonzero) [reply python/paddle/fluid/layers/nn.py:2802]: The output type is still fp16, but the parameters are all fp32. please ref: https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html. and the current kernel implementation does this too.

Comment (typhoonzero) [reply python/paddle/fluid/layers/nn.py:2802]: Well the only difficulty is that when we copy "master" parameters for mix precision, we must skip batch_norm layer's all parameters.

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (panyx0718) [reply python/paddle/fluid/layers/nn.py:2801]: Then we should add arguments to bn layer to tell bn to cast parameter to fp16.

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (panyx0718) [reply python/paddle/fluid/layers/nn.py:2802]: this is strange. why input is fp16, output is fp32? What if I want to infer type, output should by fp16 when input is fp16?

Comment (typhoonzero) [reply python/paddle/fluid/layers/nn.py:2802]: The output is fp16.

explicit is better than implicit, should add an argument to control this. Will do this.
